src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs

```diff
@@ -37,49 +38,78 @@ public Task<T> AddInvocation<T>(string connectionId, string invocationId, string
                 cancellationToken,
                 () => TryCompleteResult(connectionId, CompletionMessage.WithError(invocationId, "Canceled")));
 
-            var result = _pendingInvocations.TryAdd(invocationId, new PendingInvocation(typeof(T), connectionId, tcs, static (state, completionMessage) =>
+            // When the caller server is also the client router, Azure SignalR service won't send a ServiceMappingMessage to server.
+            // To handle this condition, CallerClientResultsManager itself should record this mapping information rather than waiting for a ServiceMappingMessage sent by service. Only in this condition, this method is called with instanceId != null.
+            var result = _pendingInvocations.TryAdd(invocationId, 
+                new PendingInvocation(
+                    typeof(T), connectionId, instanceId, tcs, 
+                    static (state, completionMessage) =>
+                    {
+                        var tcs = (TaskCompletionSourceWithCancellation<T>)state;
+                        if (completionMessage.HasResult)
+                        {
+                            tcs.SetResult((T)completionMessage.Result);
+                        }
+                        else
+                        {
+                            tcs.SetException(new Exception(completionMessage.Error));
+                        }
+                    })
+            );
+            Debug.Assert(result);
+
+            tcs.RegisterCancellation();
+
+            return tcs.Task;
+        }
+
+        public void AddServiceMapping(ServiceMappingMessage serviceMappingMessage)
+        {
+            if (_pendingInvocations.TryGetValue(serviceMappingMessage.InvocationId, out var invocation))
             {
-                var tcs = (TaskCompletionSourceWithCancellation<T>)state;
-                if (completionMessage.HasResult)
+                if (invocation.RouterInstanceId == null)
                 {
-                    tcs.SetResult((T)completionMessage.Result);
+                    invocation.RouterInstanceId = serviceMappingMessage.InstanceId;
+                    _pendingInvocations[serviceMappingMessage.InvocationId] = invocation;
```

Comment:
what if this invocationId was just removed from _pendingInvocations? Change PendingInvocation to be a reference type?